sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,582 @@
+     * This method updates the following: {@link #byteCursor}, {@link #offsets},
+     * {@link #varSizeColIndex}, and all the fields related to the validity
+     *
+     * @param offsetArr if not NULL, it will be used to save the column offsets.
+     */
+    private void initFixedWidthSection(int[] offsetArr) {
+      if (offsetArr == null) {
+        for (int i = 0; i < attributes.length; i++) {
+          advanceColCursorAndGet(i);
+        }
+      } else {
+        this.offsets = offsetArr;
+        for (int i = 0; i < attributes.length; i++) {
+          offsets[i] = advanceColCursorAndGet(i);
+        }
+      }
+      // set the validity size, and mark where it starts
+      setValidityBits(attributes.length);
+    }
+
+    /**
+     * Given the column-index, it increments the {@link #byteCursor} by the length of the datatype
+     * of that column.
+     * It takes into considerations the alignment of the column.
+     * Note that this method expects sequential increments to the column. Accessing columns in
+     * random order would result in incorrect calculations.
+     *
+     * @param ind the index of the column in the schema.
+     * @return At what point the column starts.
+     */
+    private int advanceColCursorAndGet(int ind) {
+      int res;
+      Attribute attr = attributes[ind];
+      DataType dataType = attr.dataType();
+      DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+      if (isFixedLength(dataType)) {
+        int length = rapidsType.getSizeInBytes();
+        res = alignOffset(byteCursor, length);
+        byteCursor = res + length;
+        return res;
+      }
+      if (!hasVarSizeData()) {
+        // set the index to the first variable size column.
+        varSizeColIndex = ind;
+      }
+      res = alignOffset(byteCursor, JCUDF_VAR_LENGTH_FIELD_ALIGNMENT);
+      byteCursor = res + JCUDF_VAR_LENGTH_FIELD_SIZE;
+      return res;
+    }
+
+    private void resetMeta() {
+      varSizeColIndex = attributes.length;
+      byteCursor = 0;
+      estimateRowSize = -1;
+      validityBytesOffset = 0;
+    }
+
+    private void setValidityBits(int fieldsCount) {
+      validityBytesOffset = byteCursor;
+      byteCursor += calculateBitSetWidthInBytes(fieldsCount);
+    }
+
+    /**
+     * This is used internally to estimate the memory needed by the row.
+     * The value calculated is 8-bytes aligned.
+     * It loops on variable-width columns using a moving-cursor that represents the offset
+     * of the data.
+     * Note that the calculation requires that {@link #initFixedWidthSection(int[])} has
+     * been called at some point before. Otherwise, the {@link #byteCursor} will be 0.
+     */
+    private void calculateEstimateSize() {
+      // at this point the cursor points to the beginning of the variableWidth section
+      estimateRowSize = byteCursor;
+      if (hasVarSizeData()) {
+        for (int ind = varSizeColIndex; ind < attributes.length; ind++) {
```

Comment:
I had a similar issue and I was worried about this as well. Instead of saving off the first variable-width and marching from there, I chose to build a vector of variable-width columns as I calculated the fixed-width size. Not sure how common it will be to have one variable-width column at the beginning of enough fixed-width columns for any of this to matter though.

Reply:
I agree, I considered that implementation as well.
However, I considered the following:

- Estimating the row size should be called _only once_ during the initialization of the rowToColumnarBatch.
- I was considering the memory/time-tradeoff. The above implementation allocates a java `ArrayList` on every time we want to estimate a row size. I was worried about the impact of cache performance and GC overhead

Reply:
Yes, and the number of columns is going to be small as well.